subprojects/build-cache-http/src/main/java/org/gradle/caching/http/internal/AbstractHttpBuildCacheService.java

```diff
@@ -181,6 +178,8 @@ public boolean isStreaming() {
         }
     }
 
+    protected abstract URI resolveUriForKey(URI root, BuildCacheKey key);
```

Comment:
:x: This will be a lot simpler if we use an injected `CacheKeyUriResolver` instead of inheritance.